### ISSUE TITLE
Add an option for allowing a sound played from Audio Playback Track from AnimationPlayer to play after animation has finished

### ISSUE THREAD

**Issue by imirui** [godotengine/godot-proposals#10874]
### Describe the project you are working on

I am working on a first person shooter game. I am attempting to find a method to play weapon sounds, and the main issue arises when a gun fires. The previous method (changing the stream of a specific audio player as an animationplayer key) is very time consuming, so I searched for another way. Enter: the Audio Playback track.
![image](https://github.com/user-attachments/assets/1cf3681d-a6e2-4cc5-97a8-ceef46ac9768)

### Describe the problem or limitation you are having in your project

The Audio Playback Track works fine if the sound doesn't last longer than the animation. It can play multiple at the same time, so it's great for sounds like magazine in/out, slide/bolt forward/back, etc. The problem is with firing sounds, since most firing animations last a very short duration compared to their sounds. For example, the animation of a DB Shotgun firing in my game lasts 14 frames at 24 fps, while the sound lasts until about frame 47. As soon as the animation ends, any and all Audio Playback tracks immediately stop whatever sounds they're playing. The same issue happens with Animation Playback tracks too.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The feature I am suggesting is a simple check box below the End Offset setting to allow the sound effect to continue after the animation ends.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

If enabled, the sound in the AudioStreamPlayer will continue after the animation itself has ended. Here's an example of how it would look in engine, if added.
![image](https://github.com/user-attachments/assets/3332e8f8-4361-491a-a225-f0e0dd201d9f)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Currently, this can be worked around in three methods: You can have a separate animation player for sounds which has the play method called when you want to have the sound play, however, this still does not allow the same sound to overlap. You can also have an audio stream player, with the stream changed and the playing property set to true on that frame, but it still has the same problems. The last workaround that can be used is by writing your own code to do it, even with specifying a file to play and allowing for audio overlap, at the cost of never being able to hear the audio until runtime, and the call method being very slow since you have to copy the path to the sound. Two of these methods require adding a method call track, which is a very slow track to set up, and the last requires two keys to be used.

### Is there a reason why this should be core and not an add-on in the asset library?

This should be a core addition to the engine as a way to extend the engine's functionality. It will significantly reduce time with adding sound effects, allowing for rapidly repeating sounds that last a long time like gun shots, impact sounds, and other similar sounds to be played at a specific time during an animation, using the built in Audio Playback track. It would also make the audio playback work how one would expect given the visuals, with the old way being optional if the sound should cease.

**Comment by Calinou** [comment]
To be fair, I wonder why audio playback even has to be stopped once the animation is done playing in the first place. I can't see any legitimate use cases for this, since the sound cuts off abruptly and you don't want that in a polished game. (You might want a fade-out effect, in which case you'd be animating the AudioStreamPlayer's `volume_db` property over time manually.)

**Comment by imirui** [comment]
I forgot to mention it in the proposal when I wrote it, but the same thing happens with animation players and animation playback tracks. I can understand why they would stop if you want any animations to end after a cutscene does. However, I can not understand the reasoning for it in Audio Playback Tracks since all the audio playing is put into a polyphonic stream, so it getting cut off shouldn't be an issue.

**Comment by LiTianchu** [comment]
Just encountered this issue in my project, hope there is a way to fix it.

**Comment by Octelly** [comment]
Just stumbled upon this problem myself. Are there any good workarounds for this?

**Comment by Octelly** [comment]
Sorry to be spamming an issue like this, but just tried setting the stream and playing properties manually from the animation and I completely borked my project by doing this, even managed to make Godot crash. :/

This really makes no sense to me.

**Comment by HungryHedgehog** [comment]
Agreed, got here trying to look if this feature exists. Adding this would allow gunshots to be synced to the animation but the animation still end before the sound rings out (so you can use animation length for fire rate for example)

**Comment by Calinou** [comment]
@TokageItLab Do you know how difficult this would be to address?

**Comment by TokageItLab** [comment]
I believe it needs to be stopped at least on the editor, because we have gotten issue like https://github.com/godotengine/godot/issues/71810 in the past.

BTW, I assume that it would be difficult to implement an option to keep what was played before the stop after the stop. Because AnimationPlayer no longer processes (by `_set_process(false)`) after stopping, AudioPolyphonic will remain after the audio that played before stopping is over. I think you could arbitrarily destroy them with clear_caches, but that is not very user friendly.

On the other hand, AnimationTree does not have this problem because it is processed even if the animation is not playing. So, possibly an approach that implements an option in AnimationPlayer to continue processing if there is something that played before it stopped could work.

**Comment by kaluluosi** [comment]
I think the `Continue After Animation` function does not need to take effect in the AnimationPlayer editor preview; it only needs to work during game runtime.

What we want is simply that the sound played by the sound track during game runtime does not get interrupted when the animation ends.

The editor's preview is just an extra feature and should not overly pursue consistency with the runtime.

So simply put, my suggestions are:

1. In the editor state, `Continue After Animation` does not need to be handled; keep the current behavior where the sound stops when the animation ends in the editor.
2. During game runtime, `Continue After Animation` will continue processing the unfinished audio stream after the animation ends.

This way, the risk of memory leaks in the editor is much lower.

**Comment by hsandt** [comment]
I just switched from Call Method Track + `play_sfx` method to Audio Playback Track today and I really enjoyed the preview feature so that proposal would really help with my workflow. My current hack is to extend the animation time but that modifies gameplay balance since it adds lag at the end of the enemy's action, and I'm waiting for animation to end to proceed (and otherwise it wouldn't finish playing the SFX anyway).

About the alternatives, I must say that it's easier to use the Call Method Track today (Godot 4.6) as the Inspector interface now shows an Object reference field so you can use Quick Load and drag-and-drop to find the needed asset (no crazy copying resource paths around). Still a few extra clicks to unfold Args > 0 and obviously no preview, though.

<img width="636" height="192" alt="Image" src="https://github.com/user-attachments/assets/1f23218a-1ed8-4d55-997e-fa8709dbd6bd" />

<img width="536" height="382" alt="Image" src="https://github.com/user-attachments/assets/408635ee-397d-4c87-b8ab-d5b0bcd031e0" />

The audio track preview is so useful, I may use it even if just to iterate and eventually disable it and replace it with call methods for trailing sounds...

If this proposal is implemented, we need to think about what happens when another animation is played with its own Audio Playback track using the same Audio Streamer and some Audio Clip. I assume it would interrupt the previous audio clip in this case since there would be only one Audio Streamer. Should it support blending with previous track, even if it comes from another animation? I personally don't mind having it completely replaced but many some users would like cross-animation blending. In the worst case, they could create a second Audio Streamer but that may complicate the setup.
